assumeutxo: Don't load a snapshot if it's not in the best header chain

mzumsande opened the pull request:
This was suggested by me in the discussion of #30288, which has more context.

If the snapshot is not an ancestor of the most-work header (`m_best_header`), syncing from that alternative chain leading to  `m_best_header` should be prioritised. Therefore it's not useful loading the snapshot in this situation.
If the other chain turns out to be invalid or the chain with the snapshot retrieves additional headers so that it's the most-work one again (see functional test), `m_best_header` will change and loading the snapshot will be possible again.

Because of the work required to generate a conflicting headers chain, a situation with two conflicting chains should only be possible under extreme circumstances, such as major forks.

mzumsande committed:
validation: Don't load a snapshot if it's not in the best header chain.

If the snapshot is not an ancestor of the most-work header (m_best_header),
syncing from that alternative chain should be prioritised.
Therefore don't accept loading a snapshot in this situation.

If that other chain turns out to be invalid, m_best_header
would be reset and loading the snapshot should be possible again.

Because of the work required to generate a conflicting headers chain,
this should only be possible under extreme circumstances, such as major forks.

fjahr commented on src/validation.cpp:
```diff
@@ -5676,6 +5676,9 @@ util::Result<void> ChainstateManager::ActivateSnapshot(
                           base_blockhash.ToString())};
         }
+        if (!m_best_header || m_best_header->GetAncestor(base_blockheight) != snapshot_start_block) {
+            return util::Error{_("A forked headers-chain with more work than the chain with the base block header exists. Please proceed to sync without AssumeUtxo.")};
```
nit: Not sure if users understand what "base block" is. Maybe make it "snapshot base block".

mzumsande replied:
will change to that when taking it out of draft.

fjahr reviewed (commented):
Concept ACK

mzumsande commented:
After merge of #30267: Rebased, addressed nit by fjahr, and marked as ready for review.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -218,6 +222,27 @@ def test_snapshot_block_invalidated(self, dump_output_path):
+        # create two blocks that form a longer chain
+        parent = node0.getblockhash(SNAPSHOT_BASE_HEIGHT - 1)
+        block_time = node0.getblock(node0.getbestblockhash())['time'] + 1
+        fork1 = create_block(int(parent, 16), create_coinbase(SNAPSHOT_BASE_HEIGHT - 1), block_time)
```
Huh, either I am confused or something isn't right here: The `parent` above is at height `SNAPSHOT_BASE_HEIGHT - 1` and now the child of this `parent` has a coinbase also at height `SNAPSHOT_BASE_HEIGHT - 1`?  It seems this isn't validated so it's not a huge issue though.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -218,6 +222,27 @@ def test_snapshot_block_invalidated(self, dump_output_path):
+        self.log.info("Test bitcoind should fail when the node knows the headers of another chain with more work.")
+        node0 = self.nodes[0]
+        node1 = self.nodes[1]
+        # create two blocks that form a longer chain
```
nit: Comment could be a little more detailed

```suggestion
# On node 0, create an alternative chain containing 2 new blocks, forking off the main chain at the block before the snapshot block. This simulates a longer chain than the main chain when submitting these two block headers to node 1 because it is only aware of the main chain headers up to the snapshot height.
```

mzumsande replied:
Done, in a slightly modified way. The alternative chain is just created in python, node 0 is only used to fetch the previous block it builds on.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -218,6 +222,27 @@ def test_snapshot_block_invalidated(self, dump_output_path):
+        msg = "A forked headers-chain with more work than the chain with the snapshot base block header exists. Please proceed to sync without AssumeUtxo."
+        assert_raises_rpc_error(-32603, msg, node1.loadtxoutset, dump_output_path)
+        # Cleanup: submit two more headers of the snapshot chain, so that loading the snapshot in future subtests succeeds
+        main1 = node0.getblock(node0.getblockhash(SNAPSHOT_BASE_HEIGHT + 1), 0)
```
nit: making the variables a bit more detailed could make the test a bit easier to parse

`parent` = > `parent_block_hash`
`fork1` => `fork_block1`
`main1` = `main_block1`

mzumsande replied:
took those

fjahr reviewed (commented):
tACK 29ab88dfe1c43af3620480c99cba844aa414023c

I reviewed the code and confirmed that the test fails as expected when the behavior change is not applied.

fjahr commented on test/functional/feature_assumeutxo.py:
```diff
@@ -218,6 +222,27 @@ def test_snapshot_block_invalidated(self, dump_output_path):
             node.reconsiderblock(block_hash)
 
+    def test_snapshot_not_on_most_work_chain(self, dump_output_path):
+        self.log.info("Test bitcoind should fail when the node knows the headers of another chain with more work.")
```
I think this test also actually resolves the TODOs that https://github.com/bitcoin/bitcoin/pull/29996/commits/556ff68cea1dfa6c444ef9c920dd0ce52ecc9668 aimed to resolve. I will elaborate more there.

fjahr replied:
Can remove the TODO `Valid snapshot file and snapshot block, but the block is not on the most-work chain` here now as well.

fjahr replied:
Actually also this TODO, I think these basically mean the same thing: `- TODO: Not an ancestor or a descendant of the snapshot block and has more work`

fjahr replied:
Could also mention #28648 as related in description because of this.

mzumsande replied:
Removed both TODOs.

mzumsande commented:
[29ab88d ](https://github.com/bitcoin/bitcoin/commit/29ab88dfe1c43af3620480c99cba844aa414023c)to [ab478c5](https://github.com/bitcoin/bitcoin/commit/ab478c5fa16427b496e8a93e4780770d4c270214):
Addressed feedback by @fjahr (thanks!) and rebased.

fjahr commented on src/validation.cpp:
```diff
@@ -5681,6 +5681,10 @@ util::Result<void> ChainstateManager::ActivateSnapshot(
         }
 
+        if (!m_best_header || m_best_header->GetAncestor(base_blockheight) != snapshot_start_block) {
+            return util::Error{_("A forked headers-chain with more work than the chain with the base block header exists. Please proceed to sync without AssumeUtxo.")};
```
nit: seems the "snapshot" added before is back out, still not a block for me though

mzumsande replied:
ah, sorry, am on a different computer now and forgot to update the branch... Fixed now.

fjahr reviewed (commented):
re-ACK ab478c5fa16427b496e8a93e4780770d4c270214

Confirmed only changes since last review were the rebase and addressing comments (per `git range-diff master 29ab88dfe1c43af3620480c99cba844aa414023c ab478c5fa16427b496e8a93e4780770d4c270214`).

fjahr commented:
re-ACK b715a13d11c6ca7e928b2dee66edcc8c6c30a8d4

Just addressed my latest re-nit.

alfonsoromanz reviewed (approved):
tACK b715a13d11c6ca7e928b2dee66edcc8c6c30a8d4

The test passes successfully. I also verified that the test fails with no exception raised if the patch in `src/validation.cpp` is not applied, i.e

> File "/Users/alfonso/dev/bitcoin/test/functional/feature_assumeutxo.py", line 238, in test_snapshot_not_on_most_work_chain
>     assert_raises_rpc_error(-32603, msg, node1.loadtxoutset, dump_output_path)
>   File "/Users/alfonso/dev/bitcoin/test/functional/test_framework/util.py", line 148, in assert_raises_rpc_error
>     assert try_rpc(code, message, fun, *args, **kwds), "No exception raised"
> AssertionError: No exception raised

mzumsande commented:
[b715a13 ](https://github.com/bitcoin/bitcoin/commit/b715a13d11c6ca7e928b2dee66edcc8c6c30a8d4)to [55b6d7b](https://github.com/bitcoin/bitcoin/commit/55b6d7be68a6f6c3882588ffd5b9349d885ed953): rebased

fjahr commented:
re-ACK 55b6d7be68a6f6c3882588ffd5b9349d885ed953

alfonsoromanz reviewed (approved):
Re ACK 55b6d7be68a6f6c3882588ffd5b9349d885ed953

achow101 commented:
ACK 55b6d7be68a6f6c3882588ffd5b9349d885ed953

Sjors commented:
Post merge concept ACK
